xla/service/gpu/fusions/cudnn.cc

```diff
@@ -0,0 +1,73 @@
+/* Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/gpu/fusions/cudnn.h"
+
+#include "absl/strings/escaping.h"
+#include "xla/hlo/ir/hlo_instructions.h"
+#include "xla/service/gpu/runtime/cudnn_thunk.h"
+#include "xla/service/llvm_ir/llvm_util.h"
+
+namespace xla {
+namespace gpu {
+
+absl::StatusOr<FusionEmissionResult> CuDnnFusion::Emit(
+    IrEmitterContext& ir_emitter_context, mlir::lmhlo::FusionOp fusion_op,
+    const HloFusionInstruction& fusion) const {
+#if GOOGLE_CUDA
+  if (ir_emitter_context.emit_ir_from_hlo()) {
+    VLOG(3) << fusion.ToString();
+  } else {
+    CHECK_NE(fusion_op, nullptr);
+    VLOG(3) << llvm_ir::DumpToString(fusion_op);
+  }
+
+  TF_ASSIGN_OR_RETURN(auto gpu_config,
+                      fusion.backend_config<GpuBackendConfig>());
+  if (!gpu_config.fusion_backend_config().has_cudnn_fusion_config() ||
+      !gpu_config.fusion_backend_config()
+           .cudnn_fusion_config()
+           .has_serialized_graph()) {
+    return absl::FailedPreconditionError("cuDNN fusion is not compiled.");
```

Comment:
I think we should probably move this checks to a factory function (and just have the backend config as a field). Emit is too late for this kind of error handling. Likewise the CUnescape below.

Reply:
I did not catch the idea yet. What do you mean by a factory function? Should the checks happen somewhere significantly earlier?

Reply:
I think ideally we'd have a function that returns a StatusOr<std::unique_ptr<CuDnnFusion>> so we could still fall back to some other code generation strategy. But I assume any of these checks failing indicates we have a compiler bug anyway?

Reply:
Right, missing or invalid serialized_graph means invalid IR here, most likely caused by a compiler bug. I don't see a useful case for a fallback here yet. At least for now cuDNN GEMM fusions will only be created by the GEMM fusion autotuner pass based on measurements (coming in a next PR). Having such a fusion in the IR means that it is known to be compilable and executable.

Reply:
OK, leaving as is works for me for now. Thanks!